crypto/src/bulletproofs.rs

```diff
@@ -0,0 +1,263 @@
+            point,
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl ops::Add<PedersenCommitment> for PedersenCommitment {
+    type Output = PedersenCommitment;
+
+    fn add(self, rhs: PedersenCommitment) -> PedersenCommitment {
+        PedersenCommitment {
+            point: self.point + rhs.point,
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl ops::Sub<PedersenCommitment> for PedersenCommitment {
+    type Output = PedersenCommitment;
+
+    fn sub(self, rhs: PedersenCommitment) -> PedersenCommitment {
+        PedersenCommitment {
+            point: self.point - rhs.point,
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl AsRef<[u8]> for PedersenCommitment {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.get_or_init(|| self.point.compress().to_bytes())
+    }
+}
+
+impl ToFromBytes for PedersenCommitment {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != 32 {
+            return Err(signature::Error::new());
+        }
+        let point = CompressedRistretto::from_slice(bytes);
+        let decompressed_point = point.decompress().ok_or_else(signature::Error::new)?;
+
+        Ok(PedersenCommitment {
+            point: decompressed_point,
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+impl Serialize for PedersenCommitment {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let bytes = self.as_ref();
+        serializer.serialize_bytes(bytes)
+    }
+}
+
+impl<'de> Deserialize<'de> for PedersenCommitment {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let bytes = Vec::deserialize(deserializer)?;
+        PedersenCommitment::from_bytes(&bytes[..]).map_err(|e| de::Error::custom(e.to_string()))
+    }
+}
+
+impl PartialEq for PedersenCommitment {
+    fn eq(&self, other: &Self) -> bool {
+        self.point == other.point
+    }
+}
+
+impl Eq for PedersenCommitment {}
+
+impl PartialOrd for PedersenCommitment {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_ref().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for PedersenCommitment {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_bytes().cmp(other.as_bytes())
+    }
+}
+
+///
+/// Bulletproof Range Proofs
+///
+
+#[derive(Debug)]
+pub struct BulletproofsRangeProof {
+    proof: RangeProof,
+    bytes: OnceCell<Vec<u8>>,
+}
+
+impl BulletproofsRangeProof {
+    // Prove that commited value is within the range [0, 2^64)
+    pub fn prove_bound(
+        value: u64,
+        blinding: [u8; 32],
+    ) -> Result<(PedersenCommitment, Self), signature::Error> {
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(64, 1);
+        let mut prover_transcript = Transcript::new(b"mysten");
```

Comment:
Not familiar with the `Transcript` but as far as I understood this is being used for domain separation. In this case do we want to go with `mysten` or rather the specific project (ie `narwhal`). I mean is this effectively allow the proofs to work across our products (ex assuming SUI is also using the same domain). We might (or not) want that. Apologies if my observation is totally off - not familiar with the  foundation.

Reply:
Yes, I believe this is domain separation - I'm happy to go with narwhal for this PR, but I think this entire crate is actually going to be moved to mysten-infra